semantic_router/splitters/rolling_window.py

```diff
@@ -0,0 +1,326 @@
+
+    def find_split_indices(self, similarities: List[float]) -> List[int]:
+        split_indices = []
+        for idx in range(1, len(similarities)):
+            if similarities[idx] < self.calculated_threshold:
+                split_indices.append(idx + 1)
+        return split_indices
+
+    def find_optimal_threshold(self, docs: List[str], encoded_docs: np.ndarray):
+        token_counts = [tiktoken_length(doc) for doc in docs]
+        cumulative_token_counts = np.cumsum([0] + token_counts)
+        similarity_scores = self.calculate_similarity_scores(encoded_docs)
+
+        # Analyze the distribution of similarity scores to set initial bounds
+        median_score = np.median(similarity_scores)
+        std_dev = np.std(similarity_scores)
+
+        # Set initial bounds based on median and standard deviation
+        low = max(0.0, float(median_score - std_dev))
+        high = min(1.0, float(median_score + std_dev))
+
+        iteration = 0
+        while low <= high:
+            self.calculated_threshold = (low + high) / 2
+            logger.info(
+                f"Iteration {iteration}: Trying threshold: {self.calculated_threshold}"
+            )
+            split_indices = self.find_split_indices(similarity_scores)
+
+            # Calculate the token counts for each split using the cumulative sums
+            split_token_counts = [
+                cumulative_token_counts[end] - cumulative_token_counts[start]
+                for start, end in zip(
+                    [0] + split_indices, split_indices + [len(token_counts)]
+                )
+            ]
+
+            # Calculate the median token count for the splits
+            median_tokens = np.median(split_token_counts)
+            logger.info(
+                f"Iteration {iteration}: Median tokens per split: {median_tokens}"
+            )
+            if (
+                self.min_split_tokens - self.split_tokens_tolerance
+                <= median_tokens
+                <= self.max_split_tokens + self.split_tokens_tolerance
+            ):
+                logger.info(
+                    f"Iteration {iteration}: "
+                    f"Optimal threshold {self.calculated_threshold} found "
+                    f"with median tokens ({median_tokens}) in target range "
+                    f" {self.min_split_tokens}-{self.max_split_tokens}."
+                )
+                break
+            elif median_tokens < self.min_split_tokens:
+                high = self.calculated_threshold - self.threshold_adjustment
+                logger.info(f"Iteration {iteration}: Adjusting high to {high}")
+            else:
+                low = self.calculated_threshold + self.threshold_adjustment
+                logger.info(f"Iteration {iteration}: Adjusting low to {low}")
+            iteration += 1
+
+        logger.info(f"Final optimal threshold: {self.calculated_threshold}")
+        return self.calculated_threshold
+
+    def split_documents(
```

Comment:
Should be __call__ method to align with other splitters

Reply:
no, I have __call__ method:
```
    def __call__(self, docs: List[str]) -> List[DocumentSplit]:
        if not docs:
            raise ValueError("At least one document is required for splitting.")

        if len(docs) == 1:
            token_count = tiktoken_length(docs[0])
            if token_count > self.max_split_tokens:
                logger.warning(
                    f"Single document exceeds the maximum token limit "
                    f"of {self.max_split_tokens}. "
                    "Splitting to sentences before semantically splitting."
                )
            docs = split_to_sentences(docs[0])
        encoded_docs = self.encode_documents(docs)
        self.find_optimal_threshold(docs, encoded_docs)
        similarities = self.calculate_similarity_scores(encoded_docs)
        split_indices = self.find_split_indices(similarities=similarities)
        splits = self.split_documents(docs, split_indices, similarities)
        self.plot_similarity_scores(similarities, split_indices)
        return splits
```